addons/l10n_sa_edi/models/account_edi_format.py

```diff
@@ -0,0 +1,515 @@
+            ('state_id', _('State / Country subdivision'))
+        ]
+        return self._l10n_sa_check_partner_missing_info(partner_id, fields_to_check)
+
+    def _l10n_sa_check_buyer_missing_info(self, invoice):
+        """
+            Helper function to check if ZATCA mandated partner fields are missing for the buyer
+        """
+        fields_to_check = []
+        if any(tax.l10n_sa_exemption_reason_code in ('VATEX-SA-HEA', 'VATEX-SA-EDU') for tax in
+               invoice.invoice_line_ids.filtered(lambda line: not line.display_type).tax_ids):
+            fields_to_check += [
+                ('l10n_sa_additional_identification_scheme',
+                 _('Additional Identification Scheme is required for the Buyer if tax exemption reason is either '
+                   'VATEX-SA-HEA or VATEX-SA-EDU, and its values must be NAT'), lambda v: v == 'NAT'),
+                ('l10n_sa_additional_identification_number',
+                 _('Additional Identification Number is required for commercial partners')),
+            ]
+        if not invoice._l10n_sa_is_simplified() and invoice.partner_id.country_id.code == 'SA':
+            # If the invoice is a non-foreign, Standard (B2B), the Building Number and Neighborhood are required
+            fields_to_check += [
+                ('l10n_sa_edi_building_number', _('Building Number for the Buyer is required on Standard Invoices')),
+                ('street2', _('Neighborhood for the Buyer is required on Standard Invoices')),
+            ]
+        return self._l10n_sa_check_partner_missing_info(invoice.commercial_partner_id, fields_to_check)
+
+    def _l10n_sa_edi_is_required_for_invoice(self, invoice):
+        """
+            Determine whether a ZATCA EDI document needs to be generated for a given invoice
+        """
+        return invoice.is_sale_document() and invoice.country_code == 'SA'
+
+    def _l10n_sa_get_invoice_name(self, invoice):
+        """
+            Generate the name of the invoice XML file according to ZATCA business rules:
+            Seller Vat Number (BT-31), Date (BT-2), Time (KSA-25), Invoice Number (BT-1)
+        """
+        vat = invoice.company_id.partner_id.commercial_partner_id.vat
+        invoice_number = re.sub("[^a-zA-Z0-9 -]", "-", invoice.name)
+        invoice_date = self._l10n_sa_get_zatca_datetime(invoice.l10n_sa_confirmation_datetime)
+        return '%s_%s_%s.xml' % (vat, invoice_date.strftime('%Y%m%dT%H%M%S'), invoice_number)
+
+    def _l10n_sa_post_zatca_edi(self, invoice):  # no batch ensure that there is only one invoice
+        """
+            Post invoice to ZATCA and return a dict of invoices and their success/attachment
+        """
+
+        # Chain integrity check: chain head must have been REALLY posted, and did not time out
+        chain_head = invoice.journal_id._l10n_sa_get_last_posted_invoice()
+        if chain_head and chain_head != invoice and not chain_head.l10n_sa_chain_index:
```

Comment:
```suggestion
        if chain_head and chain_head != invoice:
```
I don't think this is possible

Reply:
Done

Reply:
@jco-odoo actually there is one instance where this happens: In the case of a timeout, we reset the chain head to False so it gets recalculated in the next submission, as at this point we are still unsure if the invoice was accepted or rejected. For this reason, we should keep the check

Reply:
@mebe-odoo But it is already being checked in the _l10n_sa_get_last_posted_invoice

Reply:
@jco-odoo Updated **_l10n_sa_get_last_posted_invoice** to remove the check on the chain head index